Add nested nav styles

nabaroa opened the pull request:
Check out the demo for this branch at:
https://transferwise.github.io/public-navigation/branch/nested-nav-styles/

I duplicated the `Item` component to `ItemDropdown` for visualisation purposes, not sure if this is the best logic to show the nested nav.

By default the `.navbar-right` links show the arrow on the right:
![image](https://user-images.githubusercontent.com/1814752/36680981-3e857cc4-1b0f-11e8-9846-262ba3bbc887.png)

Adding the `dropdown-menu-center` modifier to `dropdown-menu` class the arrow is shown on the center. Here we need some JS to center the menu relative the center of the link.
![image](https://user-images.githubusercontent.com/1814752/36681067-7103cc1e-1b0f-11e8-95e2-8381be837f11.png)

![image](https://user-images.githubusercontent.com/1814752/36681522-bc85a4a4-1b10-11e8-9f68-fa7db41506d4.png)
On mobile the action to show the submenu is based on `:focus`. Here we'll need some JS as well to get a proper interaction. Second click on the button, :focus-within...

oliverviljamaa commented on dist/public-navigation.min.js:
```diff
@@ -0,0 +1,503 @@
+!(function(e, t) {
```
Hmm, how did this get in the `dist` folder? But let's not commit this.

oliverviljamaa commented on src/PublicNavigationWithoutContent/Menu/Items/ItemDropdown/ItemDropdown.js:
```diff
@@ -0,0 +1,36 @@
+const ItemDropdown = ({ translationKey }) => (
+  <li className="dropdown">
+    <button className="dropdown-button">
+      <span className="icon icon-joint-account m-r-1 visible-xs-inline-block visible-sm-inline-block" />
```
Could we leave icons for later to avoid hardcoding?

oliverviljamaa commented on src/PublicNavigationWithoutContent/Menu/Items/ItemDropdown/ItemDropdown.js:
```diff
@@ -0,0 +1,36 @@
+    <ul className="dropdown-menu dropdown-menu-center" role="menu">
+      <li>
+        <a href="#" className="active">
+          Send money
```
Let's get rid of the hardcoded values and use the `items.json` instead for now, WDYT?

oliverviljamaa commented on src/PublicNavigationWithoutContent/Menu/Items/ItemDropdown/ItemDropdown.js:
```diff
@@ -0,0 +1,36 @@
+
+import './ItemDropdown.less';
+
+const ItemDropdown = ({ translationKey }) => (
```
Technically, it's an `Item` as well, so how about `ItemWithDropdown`?

oliverviljamaa commented on src/PublicNavigationWithoutContent/Menu/Items/ItemDropdown/ItemDropdown.less:
```diff
@@ -0,0 +1,107 @@
+    &:hover {
+      .dropdown-menu {
+        display: block;
+        @media (max-width: @grid-float-breakpoint-max) {
```
Can we have the styles mobile-first and use the `@screen-*` variables?

nabaroa replied:
Yes, we can. But I used this option because we use it on bootstrap like that and tried to find a way to a better integration on it. Maybe it´s time to change it and take our own desicions? @grid-float-breakpoint-max is the variable used to point at which the navbar becomes uncollapsed and thought is was a good use for this.

oliverviljamaa commented on src/PublicNavigationWithoutContent/Menu/Items/ItemDropdown/ItemDropdown.less:
```diff
@@ -0,0 +1,107 @@
+  }
+
+  .navbar-collapse-panel {
+    padding: 0;
```
Hmm, why do we need this?

oliverviljamaa commented on src/PublicNavigationWithoutContent/Menu/Items/Items.js:
```diff
@@ -2,9 +2,11 @@ import React from 'react';
 
 const Items = ({ items }) => (
   <ul className="nav navbar-nav navbar-right">
+    {items.map(itemProps => <ItemDropdown {...itemProps} key={itemProps.translationKey} />)}
```
Could we... just have `Item`s and if they have a nested `items` object of their own, render the additional dropdown DOM inside?

nabaroa commented:
Updated
* The other one was based on <button> element for a more semantic markup, but this one is more similar of how we use the navbar items bootstrap.
* Added `active` state to show current page
* Added `.focus` state

oliverviljamaa commented:
Nice work! A few first comments.

1) When clicking on an item with subitems in the mobile menu, it redirects immediately, so I think we should remove the `link` for the dropdown parent `item`s.

2) The links on the mobile menu look a bit big, is that intended?

![image](https://user-images.githubusercontent.com/5443561/37404019-820b9a7a-2788-11e8-97f9-d752236e638a.png)

nabaroa commented:
In the designs the items on mobile have 18px, I double checked with the designers and thats correct, but told me that we can use 16px if we see it's more appropriate. Having into account that the nav and the subnav have the same indentation, 18px works for me by now. But we can check it again when we have the icons and the caret implemented.

StevePole commented on src/Navigation/Menu/Items/Item/Dropdown/Dropdown.less:
```diff
@@ -0,0 +1,92 @@
+@dropdown-arrow-width: 10px;
+
+.tw-public-navigation-menu {
+  .navbar-nav {
```
The only thing is that I think most of these styles should be in Bootstrap.  They don't seem specific to public?

oliverviljamaa replied:
Agreed, but think this can happen in steps, faster to work with it here and commit to moving them over to Bootstrap after the initial test, as highly likely we'll discover a few UI bugs.

oliverviljamaa commented:
The mobile menu dropdowns don't seem to work on mobile, is this known?

oliverviljamaa commented:
Inspecting the elements yields some interesting results (try on desktop, f.e.). For simplicity, we should only add hover on desktop (so not for the menu).

oliverviljamaa commented on src/Navigation/Menu/Items/Item/Dropdown/Dropdown.less:
```diff
@@ -0,0 +1,92 @@
+        }
+      }
+
+      &.focus-within {
```
Why do we have this again? Shouldn't the styles be the same for the other states as well?

nabaroa replied:
Agree. Rearranged.

oliverviljamaa commented on src/Navigation/Menu/Items/Item/Dropdown/Dropdown.less:
```diff
@@ -0,0 +1,92 @@
+        @media (max-width: @grid-float-breakpoint) {
+          padding: @padding-large-vertical (@spacer * 4);
+          border-bottom: 1px solid @gray-base;
+          font-size: 18px;
```
Do we have a font size variable for this? Also, where did we get this size from?

oliverviljamaa commented on src/Navigation/Menu/Items/Item/Dropdown/Dropdown.less:
```diff
@@ -0,0 +1,92 @@
+
+    .dropdown-menu {
+      @media (max-width: @grid-float-breakpoint-max) {
+        position: static;
```
Why do we need static here?

nabaroa replied:
Yes, we need it. By default the dropdown-menu has an absolute position. In bootstrap we use it this way.

oliverviljamaa commented on src/Navigation/Menu/Items/Item/Dropdown/Dropdown.less:
```diff
@@ -0,0 +1,92 @@
+  .navbar-nav {
+    border-bottom: 0;
+
+    > li {
```
For simplicity, doesn't it make sense to make a clear distinction in the different styles already on the top level? So we'd have a block for up to the breakpoint and a block from there onwards?

nabaroa replied:
I know it's not ideal, but what do you think if I leave this task of thinking to create a simpler arrangement to the moment of merging into bootstrap? 
Now in the navbar.less the styles are created in this way, declaring the element and creating the differences with max and min medias.

oliverviljamaa commented on src/items/itemObjects.js:
```diff
@@ -1,43 +1,43 @@
-    //     link: 'https://transferwise.com/our-story',
-    //   },
-    // ],
+    link: '#',
```
We can completely omit the `link` property 🙂

nabaroa replied:
If I remove the `link` property the dropdowns don´t open on mobile

oliverviljamaa commented on src/Navigation/Menu/Items/Item/Dropdown/Dropdown.less:
```diff
@@ -0,0 +1,92 @@
+  .navbar-nav {
+    border-bottom: 0;
+
+    > li {
```
What are we going to do with the carets?

oliverviljamaa commented:
Regarding the carets, I feel they miss a `m-l-1` (on mobile and desktop). And is there a reason why they're our light blue on mobile?

nabaroa commented:
Light blue carets on mobile
![image](https://user-images.githubusercontent.com/1814752/38099541-c9c8ee3a-3372-11e8-86dd-80a98190ae3b.png)

oliverviljamaa commented:
Thanks for the carets screenshot! I guess it just looks slightly off, because for now, we're using the thick carets, with the smaller lightweight ones it looks a bit better.
